### PR TITLE
Remove redundant windowproxy transplanting

### PR DESCRIPTION
**Pull request by gterzian**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23967)

### CODE REVIEW

**Comment by gterzian** [comment]
@asajeffrey The call to `self.window_proxy().set_currently_active(self);` inside `window.resume` seems redundant in many cases:

1. When a new page is loaded, and we create a new windowproxy. See https://github.com/servo/servo/blob/d7433c9a6517f3e0596d9a227c6fabe167604a19/components/script/script_thread.rs#L3270
2. When a new page is loaded, and we re-use a windowproxy, since we will already do the transplant at https://github.com/servo/servo/blob/d7433c9a6517f3e0596d9a227c6fabe167604a19/components/script/script_thread.rs#L3021

So I think the call is only necessary when we set the activity of a document to full-active, and there is no in progress load for that document(essentially I guess that's when re re-use a doc in the bfcache, and just set it to fully active). 

In that case, the call is done inside `document.set_activity`, via a call to `window.resume`. I think it's probably clearer to do it as I've changed it here, and to remove the call inside `window.resume`, since most cases when we resume a window we do not need to transplant the windowproxy.

**Comment by asajeffrey** [comment]
This seems like a dangerous way to fix the issue, calling `resume` should make the document active. I'm concerned that with this change, we'd get weird race conditions about things like nested documents A with iframe B with iframe C, if C is resumed then B is resumed, since this should result in C being fully active but I'm not sure it will with this revision.

Is the case you're worried about one where we're transplanting a window proxy even though it's already in the right realm? If so, can we deal with this by adding a check to `set_window` to exit early if the proxy is already in the same realm as the window?

**Comment by gterzian** [comment]
> Is the case you're worried about one where we're transplanting a window proxy even though it's already in the right realm?

What I've noticed is that when we load a new url in an existing browsing-context, we hit the branch at https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/script_thread.rs#L3021

Then later we hit it again at https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/dom/window.rs#L1949

So we simply do the transplant twice, as far as I can tell. 

In the other cases, we do the transplant only once as part of `window.resume`, and I still wonder whether it's not redundant, since in those other cases we actually create a windowproxy using the window, as in `WindowProxy::new(&window, ..)`. Do we then still need to do a `set_window` after that? 

The only other place in the code I could find that uses `window.resume` is  at https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/dom/document.rs#L507, 

which is called from https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/script_thread.rs#L2410

That case seems necessary, since it deals with an existing window that becomes fully-active inside an existing browsing context. 

Could we catch the redundancies inside https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/dom/windowproxy.rs#L605

by checking the pipeline id of the window against `self.currently_active`? Yes we could because in those cases they're already the same, and then I assume the call to `set_window` is redundant?

It might still be better to only call `set_currently_active` when it's necessary, as opposed to making it idempotent?

> I'm concerned that with this change, we'd get weird race conditions about things like nested documents A with iframe B with iframe C, if C is resumed then B is resumed, since this should result in C being fully active but I'm not sure it will with this revision.

There might be something I'm missing, and from reading the code it seems the above mentioned cases provide an exhaustive list of when `set_window` is called, so if the redundancies are correctly identified it seems we should not get weird side-effects. 

Iframes are dealt as part of https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/script_thread.rs#L3024

and it seems that they get their own newly created windowproxy each time, hence the subsequent call to `set_window` that happens as part of `window.resume` would seem redundant.

**Comment by asajeffrey** [comment]
I guess I'm worried that this change removes the transplanting from `resume` but not `suspend`, so they're not inverses. I see what you mean though, it is odd that `local_window_proxy` has a side-effect of setting the active document.

**Comment by gterzian** [comment]
Ok, should I simply check the pipeline of the window against the currently active one, and return early, inside https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/dom/windowproxy.rs#L605

That way we keep the current structure, and we just make "redundant" calls idempotent?

If `self.currently_active` is already the same as the pipeline id of the `window` passed as an argument, we never have to actually call `set_window`, or is there something more going on?

**Comment by gterzian** [comment]
By the way, I was also surprised to see that a call to `suspend` always sets a `DissimilarOriginWindow` as the current window on the proxy... 

See https://github.com/servo/servo/blob/c47575535ceb798ae3be60792c98bdd0a33f4415/components/script/dom/windowproxy.rs#L613

**Comment by asajeffrey** [comment]
Yeah, that is odd. I think it's because if the replacement window is similar-origin, then the constellation will send another message activating it. But it does mean there's a transient state where it appears to be a dissimilar-origin window, possibly resulting in spurious security errors.

**Comment by gterzian** [comment]
Ok, the latest commit provides a fix that doesn't change the overall code stucture, ready for another review @asajeffrey

File: components/script/dom/windowproxy.rs [hunk 1]
```diff
@@ -602,12 +602,23 @@ impl WindowProxy {
     }
 
     pub fn set_currently_active(&self, window: &Window) {
-        let globalscope = window.upcast();
+        let globalscope = window.upcast::<GlobalScope>();
+        let dest_pipeline_id = globalscope.pipeline_id();
+        if let Some(pipeline_id) = self.currently_active() {
+            if pipeline_id == dest_pipeline_id {
+                return warn!(
```
Comment (asajeffrey) [inline components/script/dom/windowproxy.rs]: Make this a `debug!` rather than a `warn!`?

**Review by asajeffrey — APPROVED** [review]
A possible niggle about log levels, but otherwise LGTM.

File: components/script/dom/windowproxy.rs (hunk 1 above)
Comment (gterzian) [reply components/script/dom/windowproxy.rs]: Ok, done.

**Pull request merged**